deps/rabbit/src/rabbit_queue_member_eval.erl

```diff
@@ -0,0 +1,184 @@
+    {ok, #state{timer_ref = Ref,
+                default_interval = DefaultInterval,
+                short_interval = ShortInterval,
+                target_group_size = TargetGroupSize}}.
+
+handle_call(_Request, _From, State) ->
+    {reply, ok, State}.
+
+handle_cast(nodechange, #state{timer_ref = OldRef,
+                               short_interval = Time} = State) ->
+    _ = erlang:cancel_timer(OldRef),
+    Ref = erlang:send_after(Time, self(), ?EVAL_MSG),
+    {noreply, State#state{timer_ref = Ref}};
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info(?EVAL_MSG, #state{default_interval = DefaultInterval,
+                              short_interval = ShortInterval,
+                              target_group_size = TargetSize} = State) ->
+    Res = eval_quroum_queue_membership(TargetSize),
+    NewTimeout = case Res of
+                     noop ->
+                         DefaultInterval;
+                     _ ->
+                         ShortInterval
+                 end,
+    Ref = erlang:send_after(NewTimeout, self(), ?EVAL_MSG),
+    {noreply, State#state{timer_ref = Ref}};
+handle_info(_Info, State) ->
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%%----------------------------------------------------------------------------
+%% Internal functions
+%%----------------------------------------------------------------------------
+
+eval_quroum_queue_membership(TargetSize) ->
+    LocalLeaders = rabbit_amqqueue:list_local_leaders(),
+    ExpectedNodes = rabbit_nodes:list_members(),
+    Running = rabbit_nodes:list_running(),
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize, noop).
+
+eval_quorum_members(_ExpectedNodes, _Running, [], _TargetSize, Result) ->
+    Result;
+eval_quorum_members(ExpectedNodes, Running, [Q | LocalLeaders], TargetSize, OldResult) ->
+    {ok, Members, {_, LeaderNode}} = ra:members(amqqueue:get_pid(Q)),
+
+    %% Check if Leader is indeed this node
+    Result = case LeaderNode =:= node() of
+                 true ->
+                     MemberNodes = [Node || {_, Node} <- Members],
+                     Remove = MemberNodes -- ExpectedNodes,
+                     case Remove of
+                         [] ->
+                             add_member(Q, Running, MemberNodes, TargetSize);
+                         _ ->
+                             remove_members(Q, Remove)
+                     end;
+                 false ->
+                     noop
+             end,
+    eval_quorum_members(ExpectedNodes, Running, LocalLeaders, TargetSize,
+                        update_result(OldResult, Result)).
+
+add_member(Q, Running, MemberNodes, TargetSize) ->
+    New = Running -- MemberNodes,
+    case should_add_node(MemberNodes, New, Q, TargetSize) of
+        true ->
+            %% Check current members are up to date
```

Comment:
I think I nice log message here, maybe at info level but at least at debug that describes the decision that was made and why.